### PR TITLE
feat: better onboarding process

### LINKED ISSUE #3: Generate Default Organization Configuration App
We should have a simple page that will help new partners generate a basic UbiquiBot organization config with our recommended default settings. 

They should be able to login with GitHub and then the bot can handle the rest, which would include creating the `ubiquibot-config` repository and adding a `ubiquibot-config/.github/ubiquibot-config.yml` file. 

As a side note, I think it makes sense for us to rename the repository to `.ubiquibot-config` to clearly indicate to our partners that it is meant to be a private repository. I'm not sure how far along we will be in updating all of our infrastructure to accomodate this change, but keep this in mind when you start this task. 

> > github's app installation token

> No this can be generated with Octokit. We do that right now in the kernel. We can store the "app private key" in our backend/secrets, as well as the "app id". The user just needs to pass in their installation ID for us to authenticate under their installation

> https://github.com/ubiquity/ubiquibot-kernel/blob/af8fe436648aa6012ea885582f5e45c1e97f1bb0/src/github/github-event-handler.ts#L33-L39

> @whilefoo and I have been working with this code and the authentication definitely works.

> ---

> I just realized that we can probably handle this automatically for the user. We can make a "Login with GitHub" button so that we know what organizations they are associated with, then we can look up any overlap with all the organizations that the bot is installed to and then automatically identify the installation ID. 

> More information here: https://chat.openai.com/c/70e59735-9fde-4c97-814b-e56164a59913

_Originally posted by @pavlovcik in https://github.com/ubiquity/pay.ubq.fi/issues/65#issuecomment-1945702584_

### PR DESCRIPTION
**Pull request by gentlementlegen**
Resolves #3

### CODE REVIEW

**Comment by gentlementlegen** [comment]
## What would be needed for the deployment to work
`FRONTEND_URL` should be added to the env, it corresponds to the redirect url on a successful login attempt. The Cypress tests fail because they use the files in the org, not the ones in my pull request that are up to date.

## What are the changes
The old system of entering a PAT has been changed to use OAuth login instead. for user convenience. It allows to directly login with a click, and also to list the available repositories instead of having to manually add the repository name.

## What are the drawbacks
Using the OAuth token restricts the actions that can be performed by Octokit. One of them is to set the repositories to the Ubiquibot App, as per described here https://docs.github.com/en/rest/apps/installations?apiVersion=2022-11-28#add-a-repository-to-an-app-installation
```
This endpoint does not work with GitHub App user access tokens, GitHub App installation access tokens, or fine-grained personal access tokens.
```
I don't think it is crucial, since by default apps have access to all the repositories. Yet, if we would like to inform the user about this, I thought we could display a link with a reminder like "Don't forget to check the permissions for your app <url to the repo settings>". Please let me know if anything comes to mind and I will add it.
Edit: after talking with @pavlovcik it seems that this step is not required since the app already has the authorization needed, so we should be good to go.

Screenshot of the new workflow
image

File: build/esbuild-build.ts
Comment (0x4007) [inline build/esbuild-build.ts]: Compiler directives are not a good sign. What's this

Comment (gentlementlegen) [reply build/esbuild-build.ts]: Was because the `in` is forbidden but was safe to use as it is just a Record, can change to a `forEach`

Comment (0x4007) [reply build/esbuild-build.ts]: They are there for a reason so you should change to for each if that's the case.

File: build/esbuild-build.ts
Comment (0x4007) [inline build/esbuild-build.ts]: This whole build file is outdated. Can you take the latest from pay.ubq.fi

Comment (gentlementlegen) [reply build/esbuild-build.ts]: It is now up to date. Note that `pay.ubq.fi` is also using a `for .. in` loop, might wanna update that too

File: tsconfig.json
Comment (0x4007) [inline tsconfig.json:48]: What's this for?

File: static/scripts/onboarding/onboarding.ts
Comment (0x4007) [inline static/scripts/onboarding/onboarding.ts]:
```suggestion
    const installations = appInstallations.installations.filter((installation) => installation.app_id === APP_ID);
```

You should never shorthand identifiers. This is to reduce cognitive overhead for other developers.

File: static/scripts/onboarding/github-login-button.ts
Comment (0x4007) [inline static/scripts/onboarding/github-login-button.ts]:
```suggestion
  setButton.disabled = false
```

Is the correct way to do this. `setButton` should be casted as `HTMLButtonElement`

File: static/scripts/onboarding/github-login-button.ts
Comment (0x4007) [inline static/scripts/onboarding/github-login-button.ts]:
```suggestion
  const setBtn = document.getElementById("setBtn") as HTMLButtonElement;
```

You can cast the others above as well.

File: static/scripts/onboarding/github-login-button.ts
Comment (0x4007) [inline static/scripts/onboarding/github-login-button.ts]: Perhaps it makes sense to break off these large interfaces into separate files.

**Review by 0x4007 — COMMENTED** [review]
Sorry just now had a chance to review on my computer. Nitpicks but ready to merge after those are addressed thanks.

File: static/scripts/onboarding/onboarding.ts
Comment (0x4007) [reply static/scripts/onboarding/onboarding.ts]: The only exceptions I can think of are "repo" and "org" because octokit uses that terminology

File: tsconfig.json
Comment (gentlementlegen) [reply tsconfig.json:48]: In the `esbuild` that I copied from `pay.ubq` there is an import of a `js` file
```TypeScript
import extraRpcs from "../lib/chainlist/constants/extraRpcs";
```
doesn't work without these

File: static/scripts/onboarding/onboarding.ts
Comment (gentlementlegen) [reply static/scripts/onboarding/onboarding.ts]: Totally agree, although I am not the one who wrote these, changing now!

**Review by 0x4007 — APPROVED** [review]

**Comment by 0x4007** [comment]
All looking good thanks. When you implement requested changes you should resolve each of the conversations. You can use the conversations as your checklist.

**Comment by rndquu** [comment]
This PR shouldn't have been merged with the failing [build](https://github.com/ubiquity/onboard.ubq.fi/actions/runs/8244511034/job/22546852565)

@FernandVEYRIER Where should we set the `FRONTEND_URL` env variable, [here](https://github.com/ubiquity/onboard.ubq.fi/blob/1579b6e958efed8d3c83533e3b6de2e86b79d370/.github/workflows/build.yml#L30)? And what value should we use?

**Comment by 0x4007** [comment]
> This PR shouldn't have been merged with the failing [build](https://github.com/ubiquity/onboard.ubq.fi/actions/runs/8244511034/job/22546852565)
> 
> @FernandVEYRIER Where should we set the `FRONTEND_URL` env variable, [here](https://github.com/ubiquity/onboard.ubq.fi/blob/1579b6e958efed8d3c83533e3b6de2e86b79d370/.github/workflows/build.yml#L30)? And what value should we use?

We had a discussion on this but I suppose it might have been more appropriate to do so here in the comments. Basically I got a bit confused because it should be tested on a deployed cloudflare URL from the current commit. But this means that we need to take the Cloudflare URL as the output of the continuous deploys and then pass it into Cypress to test.

Alternatively, if I'm understanding correctly, we can just build and run in GitHub Actions and use `localhost:8080`

@FernandVEYRIER can clarify

**Comment by gentlementlegen** [comment]
@rndquu The build was failing because to preserve the `secrets` we have to trigger pull requests on the context of the target repo and not to one opening the PR, so the workflow was actually not up to date. And yes for the local build / Cypress testing `localhost:8080` can be used. For the production, the deployment url (the url where the frontend is deployed) should be used.

**Comment by molecula451** [comment]
makes more sense now if we're going to refactor these sites to a specific library/framework static site that we just start on that issue and decrease the feat addition, or maybe work both in parallel

**Comment by gentlementlegen** [comment]
So currently we should just default `FRONTEND_URL` to the currently deployed url for that page.

**Comment by rndquu** [comment]
I'm 99% sure this ([one](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml#L4), [two](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml#L15)) is [vulnerable](https://securitylab.github.com/advisories/GHSL-2020-241-getsentry-sentry-workflow/) to organization secrets exfiltrating

**Comment by gentlementlegen** [comment]
Is this still an open issue, there was no patch whatsoever?

**Comment by rndquu** [comment]
@FernandVEYRIER 

As far as I understand [cypress](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml) and [knip](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/knip.yml) workflows don't work as expected right now. When `pull_request_target` is used it means that the workflow runs in the context of the base branch (`ubiquity/development`) hence it has access to organization secrets but [here](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml#L15) it checks out the code in the `ubiquity/development` branch, not the one in the PR. So right now both  [cypress](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml) and [knip](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/knip.yml) workflows are basically running only against the `ubiquity/development` branch on all PRs.

If we want to run cypress and knip workflows against the PR branch then we need to modify the checkout action this way:
```
    steps:
    - uses: actions/checkout@v2
      with:
        ref: ${{ github.event.pull_request.head.sha }}
```

But the above code allows malicious actors to exfiltrate github access token, organization secrets and all this sort of stuff [here](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml#L19) or [here](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/knip.yml#L22) (more [info](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)).

TBH I haven't yet seen a case when `pull_request_target` should really be used.

**Comment by rndquu** [comment]
> So currently we should just default `FRONTEND_URL` to the currently deployed url for that page.

So we need to:
1. Fetch deployment URL for a PR [here](https://github.com/ubiquity/onboard.ubq.fi/blob/b4f41babfd0898cb1ce6a2a1885a186844a686ab/.github/workflows/build.yml#L35)
2. Pass the deployment URL [here](https://github.com/ubiquity/onboard.ubq.fi/blob/b4f41babfd0898cb1ce6a2a1885a186844a686ab/.github/workflows/cypress-testing.yml#L25)

Right?

**Comment by gentlementlegen** [comment]
> @FernandVEYRIER
> 
> As far as I understand [cypress](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml) and [knip](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/knip.yml) workflows don't work as expected right now. When `pull_request_target` is used it means that the workflow runs in the context of the base branch (`ubiquity/development`) hence it has access to organization secrets but [here](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml#L15) it checks out the code in the `ubiquity/development` branch, not the one in the PR. So right now both [cypress](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml) and [knip](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/knip.yml) workflows are basically running only against the `ubiquity/development` branch on all PRs.
> 
> If we want to run cypress and knip workflows against the PR branch then we need to modify the checkout action this way:
> 
> ```
>     steps:
>     - uses: actions/checkout@v2
>       with:
>         ref: ${{ github.event.pull_request.head.sha }}
> ```
> 
> But the above code allows malicious actors to exfiltrate github access token, organization secrets and all this sort of stuff [here](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/cypress-testing.yml#L19) or [here](https://github.com/FernandVEYRIER/onboard.ubq.fi/blob/2cc6281088c12526d47936ef3ce4b1ad591565b2/.github/workflows/knip.yml#L22) (more [info](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)).
> 
> TBH I haven't yet seen a case when `pull_request_target` should really be used.

Makes sense. The only reason to use the `pull_request_access` was to pass the secrets indeed. Since we review the pull-requests before merging them, wouldn't it be okay to still use it? Not everyone can merge them in the repo so it's not like random malicious code can get in. It is still a security issue ofc, but if we can't use it that way then I guess we can have these workflows working properly. Otherwise, let's change when they are run.

**Comment by gentlementlegen** [comment]
> > So currently we should just default `FRONTEND_URL` to the currently deployed url for that page.
> 
> So we need to:
> 
> 1. Fetch deployment URL for a PR [here](https://github.com/ubiquity/onboard.ubq.fi/blob/b4f41babfd0898cb1ce6a2a1885a186844a686ab/.github/workflows/build.yml#L35)
> 2. Pass the deployment URL [here](https://github.com/ubiquity/onboard.ubq.fi/blob/b4f41babfd0898cb1ce6a2a1885a186844a686ab/.github/workflows/cypress-testing.yml#L25)
> 
> Right?

We can make it simple right now and just put the frontend URL of the production deployment. But yes that would be the correct behavior to be able to test for each deployment. Currently Supabase is set to allow redirects to `https://**.ubq.fi`

**Comment by rndquu** [comment]
> > > So currently we should just default `FRONTEND_URL` to the currently deployed url for that page.
> > 
> > 
> > So we need to:
> > 
> > 1. Fetch deployment URL for a PR [here](https://github.com/ubiquity/onboard.ubq.fi/blob/b4f41babfd0898cb1ce6a2a1885a186844a686ab/.github/workflows/build.yml#L35)
> > 2. Pass the deployment URL [here](https://github.com/ubiquity/onboard.ubq.fi/blob/b4f41babfd0898cb1ce6a2a1885a186844a686ab/.github/workflows/cypress-testing.yml#L25)
> > 
> > Right?
> 
> We can make it simple right now and just put the frontend URL of the production deployment. But yes that would be the correct behavior to be able to test for each deployment. Currently Supabase is set to allow redirects to `https://**.ubq.fi`

https://github.com/ubiquity/cloudflare-deploy-action/issues/4

**Comment by rndquu** [comment]
> Since we review the pull-requests before merging them, wouldn't it be okay to still use it?

It doesn't matter if we review it or not. Once this change is made any PR will be able to run code with access to organization env variables:
```
    on: pull_request_target

    steps:
    - uses: actions/checkout@v2
      with:
        ref: ${{ github.event.pull_request.head.sha }}
```

**Pull request merged**
